File: 2021/01/1.cc
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iterator>
#include <numeric>
#include <string>
#include <utility>
#include <vector>

using namespace std;

ssize_t increases(vector<ssize_t> const& v) {
    vector<ssize_t> diff;
    adjacent_difference(v.begin(), v.end(), back_inserter(diff));

    return accumulate(diff.begin(), diff.end(), 0,
                      [](ssize_t acc, ssize_t d) { return acc + (d > 0); }) -
           1;
}

ssize_t part_1(vector<ssize_t> const& lines) { return increases(lines); }

ssize_t part_2(vector<ssize_t> const& lines) {
    vector<ssize_t> windows(lines.size() - 2);
    for (size_t i = 2; i < lines.size(); i++)
        windows[i - 2] = lines[i - 2] + lines[i - 1] + lines[i];
    return increases(windows);
}

int main() {
    ifstream        in { "data" };
    vector<ssize_t> lines;

    copy(istream_iterator<ssize_t> { in }, istream_iterator<ssize_t> {}, back_inserter(lines));

    printf("Part 1: %zu.\nPart 2: %zu.\n", part_1(lines), part_2(lines));

    return 0;
}

File: 2021/01/1.rs
fn increases(i: &[usize]) -> usize {
    i.windows(2).map(|w| (w[1] > w[0]) as usize).sum()
}

fn main() {
    const INPUT: &'static str = include_str!("data");
    let input: Vec<usize> = INPUT.lines().flat_map(str::parse).collect();

    let part_1 = increases(&input);
    let part_2 = increases(&input.windows(3).map(|w| w.iter().sum()).collect::<Vec<usize>>());

    println!("Part 1: {}.\nPart 2: {}.", part_1, part_2);
}
